Add support for SAML ECP.

jdennis opened the pull request:
The modifications in this commit address the changes necessary to
support the SP component of SAML ECP. The Lasso library needs
additional modifications before SAML ECP will be fully functional,
those fixes have been submitted to upstream Lasso, mod_auth_mellon
will continue to operate correctly without the Lasso upgrade, it just
won't properly support ECP without the Lasso fixes.

Below are the major logical changes in the commit and the rationale
behind them.

* Allow compilation against older versions of Lasso by conditionally
  compiling.

  Add the following CPP symbols set by configure:

  * HAVE_ECP
  * HAVE_LASSO_UTILS_H

* Add lasso_compat.h

  If we can't include lasso utils.h than pull in our own
  local definitions so we can use some of the valuable
  utilities.

* Add ECP specific documentation file

  Documentation specific to ECP is now contained in ECP.rst
  (using reStructuredText formatting). Information on general ECP
  concepts, mod_auth_mellon user information, and internal
  mod_auth_mellon coding issues are covered.

* Add am_get_is_passive() utility

  Add utilty to get the IsPassive flag so it can be called
  from multiple places.

* Add am_validate_paos_header() utility

  This utility routine validates the PAOS HTTP header. It is used
  in conjunction with am_header_has_media_type() to determine if a
  client is ECP capable.

* Add utility function am_header_has_media_type() to check if an HTTP
  Accept header includes a specific media type. This is necessary
  because the SP detects an ECP client by the presence of a
  application/vnd.paos+xml media type in the Accept
  header. Unfortunately neither Apache nor mod_auth_mellon already had
  a function to check Accept media types so this was custom written
  and added to mod_auth_mellon.

* Add utility function am_get_assertion_consumer_service_by_binding()
  because Lasso does not expose that in it's public API. It's
  necessary to get the URL of the PAOS AssertionConsumerService.

* Add MellonECPSendIDPList config option

  This option controls whether to include a list of IDP's when
  sending an ECP PAOS <AuthnRequest> message to an ECP client.

* We need to do some bookkeeping during the processing of a
  request. Some Apache modules call this "adding a
  note". mod_auth_mellon was already doing this but because it only
  needed to track one value (the cookie value) took a shortcut and
  stuffed the cookie value into the per module request slot rather
  than defining a struct that could hold a variety of per-request
  values. To accommodate multiple per request bookkeeping values we
  define a new struct, am_req_cfg_rec, that holds the previously used
  cookie value and adds a new ECP specific value. This struct is now
  the bookkeeping data item attached to each request. To support the
  new am_req_cfg_rec struct the am_get_req_cfg macro was added (mirrors
  the existing am_get_srv_cfg, am_get_mod_cfg and am_get_dir_cfg
  macros). The am_create_request() Apache hook was added to
  initialize the am_req_cfg_rec at the beginning of the request
  pipeline.

* A new endpoint was added to handle PAOS responses from the ECP
  client. The endpoint is called "paosResponse" and lives along side
  of the existing endpoints (e.g. postResponse, artifactResponse,
  metadata, auth, logout, etc.). The new endpoint is handled by
  am_handle_paos_reply(). The metadata generation implemented in
  am_generate_metadata() was augmented to add the paosResponse
  endpoint and bind it to the SAML2 PAOS binding.

* am_handle_reply_common() was being called by am_handle_post_reply()
  and am_handle_artifact_reply() because replies share a fair amount
  of common logic. The new am_handle_paos_reply() also needs to
  utilize the same common logic in am_handle_reply_common() but ECP
  has slightly different behavior that has to be accounted for. With
  ECP there is no SP generated cookie because the SP did not initiate
  the process and has no state to track. Also the RelayState is
  optional with ECP and is carried in the PAOS header as opposed to an
  HTTP query/post parameter. The boolean flag is_paos was added as a
  parameter to am_handle_reply_common() so as to be able to
  distinguish between the PAOS and non-PAOS logic.

* Add PAOS AssertionConsumerService to automatically generated metadata.
  Note, am_get_assertion_consumer_service_by_binding() should be able
  to locate this endpoint.

* Refactor code to send <AuthnRequest>, now also supports PAOS

  The creation and initialization of a LassoLogin object is different
  for the ECP case. We want to share as much common code as possible,
  the following refactoring was done to achieve that goal.

  The function am_send_authn_request() was removed and it's logic
  moved to am_init_authn_request_common(),
  am_send_login_authn_request() and
  am_set_authn_request_content(). This allows the logic used to create
  and initialize a LassoLogin object to be shared between the PAOS and
  non-PAOS cases. am_send_paos_authn_request() also calls
  am_init_authn_request_common() and
  am_set_authn_request_content(). The function
  am_set_authn_request_content() replaces the logic at the end of
  am_send_authn_request(), it is responsible for setting the HTTP
  headers and body content based on the LassoLogin.

Signed-off-by: John Dennis <jdennis@redhat.com>

olavmrk commented:
Hi,

I have now looked through the changes here. Mostly just minor things and suggestions to make the code more readable and maintainable. A couple of them I think are bugs though (the `default_ecp_send_idplist` variable and the `PASO` string constant).

Regarding merging this, if I understand correctly, the required changes in the Lasso library are not in the repository yet. I think I will at lest hold off merging this until they are in the master branch, so that it is possible to make a quick build with the latest version of Lasso to make sure it all works. Ideally I would like to wait until it it hits an acutal release, so that one can say something like "To use the ECP functionality, you need version x.y.z of the Lasso library".

olavmrk commented:
Hi,

just want to let you know that I am leaving for my summer vacation in a couple of hours, so I won't be able to follow up on this pull request until the beginning of August.

jdennis commented:
On 07/10/2015 05:49 AM, Olav Morken wrote:
>
> Hi,
>
> just want to let you know that I am leaving for my summer vacation in 
> a couple of hours, so I won't be able to follow up on this pull 
> request until the beginning of August.
>
>

Wishing you a wonderful summer vacation Olav. Enjoy!

I have addressed each of your comments and made fixes (one commit per 
issue) that I've pushed to the ecp branch. Provided you're OK with the 
changes my suggestion would be to squash everything into one commit for 
clarity. I updated the original commit message to reflect the review 
changes but I can't amend the commit message without incurring a 
non-fastforward error. I can send you the updated commit message when 
you're ready.

Talk to you when you get back, hopefully well rested.

John

olavmrk commented:
Hi,

I'm back at work, and have looked through your changes. I'm happy with how the changes look now, so if you would squash the changes together and update the commit message that would be great.

jdennis commented:
On 08/03/2015 09:35 AM, Olav Morken wrote:
>
> Hi,
>
> I'm back at work, and have looked through your changes. I'm happy with 
> how the changes look now, so if you would squash the changes together 
> and update the commit message that would be great.
>
>
Welcome back Olav!

Thank you. I will prepare a squashed patch and update the commit message.

John

olavmrk commented:
Hi, I merged this pull request now, but discovered that the patches weren't squashed. I therefore did some ugly history rewriting in Git to remove it.

Could you send an updated patch?

olavmrk commented:
Could you also squash in something like this in the patch?

```
diff --git a/auth_mellon_handler.c b/auth_mellon_handler.c
index d9a11d5..805661a 100644
--- a/auth_mellon_handler.c
+++ b/auth_mellon_handler.c
@@ -3509,7 +3509,6 @@ int am_auth_mellon_user(request_rec *r)
 int am_check_uid(request_rec *r)
 {
     am_dir_cfg_rec *dir = am_get_dir_cfg(r);
-    am_req_cfg_rec *req_cfg = am_get_req_cfg(r);
     am_cache_entry_t *session;
     int return_code = HTTP_UNAUTHORIZED;
 
@@ -3527,6 +3526,7 @@ int am_check_uid(request_rec *r)
     }
 
 #ifdef HAVE_ECP
+    am_req_cfg_rec *req_cfg = am_get_req_cfg(r);
     if (req_cfg->ecp_authn_req) {
         ap_log_rerror(APLOG_MARK, APLOG_DEBUG, 0, r,
                       "am_check_uid is performing ECP authn request flow");
```

It silences a warning about an unused variable.

jdennis commented:
O.K. all the commits have been squashed and the unreferenced variable 
warning fixed. The ecp branch was rebased against upstream origin/master.

Hopefully everything is in good shape.

Thanks,

John

olavmrk commented:
Hi,
unfortunately I didn't get the time to look at this today, and now I am heading for a one-week vacation. The week afterwards, there is some traveling, so it may be until Thursday 17. September before I can look at it again. Sorry about that.

thijskh commented:
@jdennis I don't see the updated commits in this PR? Or am I overlooking something?

jdennis commented:
On 09/04/2015 10:01 AM, Thijs Kinkhorst wrote:
>
> @jdennis <https://github.com/jdennis> I don't see the updated commits 
> in this PR? Or am I overlooking something?
>
>

It was probably my mistake, I pushed the squashed commit to the jdennis 
fork in the ecp branch but I forget to issue a pull request, which I 
just did now, my bad I'm sorry.

commit:
6c1012e202d52742c4584fe43dc701d97b0e6fd6

FWIW, Olav asked me to squash all the previous split out commits into 
one single commit.

John
